honeybee/face.py

```diff
@@ -1,158 +1,875 @@
-            vertices: A flattened list of 3 or more vertices as (x, y, z).
-            face_type: Face type (e.g. wall, floor).
+    def horizontal_orientation(self, north_vector=Vector2D(0, 1)):
+        """A number between 0 and 360 for the orientation of the face in degrees.
+
+        0 = North, 90 = East, 180 = South, 270 = West
+
+        Args:
+            north_vector: A ladybug_geometry Vector2D for the north direction.
+                Default is the Y-axis (0, 1).
         """
-        geometry = Face3D([Point3D(*v) for v in vertices])
-        return cls(name, geometry, face_type, boundary_condition)
+        return math.degrees(
+            north_vector.angle_clockwise(Vector2D(self.normal.x, self.normal.y)))
 
-    @property
-    def properties(self):
-        """Face properties.
+    def cardinal_direction(self, north_vector=Vector2D(0, 1)):
+        """Text description for the cardinal direction that the face is pointing.
+
+        Will be one of the following: ('North', 'East', 'South', 'West')
+
+        Args:
+            north_vector: A ladybug_geometry Vector2D for the north direction.
+                Default is the Y-axis (0, 1).
+        """
+        orient = self.horizontal_orientation(north_vector)
+        if orient <= 45 or orient > 315:
+            return 'North'
+        elif orient <= 135:
+            return 'East'
+        elif orient <= 225:
+            return 'South'
+        else:
+            return 'West'
+
+    def clear_sub_faces(self):
+        """Remove all apertures and doors from the face."""
+        self.clear_apertures()
+        self.clear_doors()
+
+    def clear_apertures(self):
+        """Remove all apertures from the face."""
+        for aperture in self._apertures:
+            aperture._parent = None
+        self._apertures = []
+        self._punched_geometry = None  # reset so that it can be re-computed
+
+    def clear_doors(self):
+        """Remove all doors from the face."""
+        for door in self._apertures:
+            door._parent = None
+        self._doors = []
+        self._punched_geometry = None  # reset so that it can be re-computed
+
+    def add_aperture(self, aperture):
+        """Add an Aperture to this face.
+
+        This method does not check the co-planarity between this Face and the
+        Aperture or whether the Aperture has all vertices within the boundary of
+        this Face. To check this, the Face3D.is_sub_face() method can be used
+        with the Aperture and Face geometry before using this method or the
+        are_sub_faces_valid() method can be used afterwards.
+
+        Args:
+            aperture: An Aperture to add to this face.
+        """
+        assert isinstance(aperture, Aperture), \
+            'Expected Aperture. Got {}.'.format(type(aperture))
+        self._acceptable_sub_face_check(Aperture)
+        aperture._parent = self
+        if self.normal.angle(aperture.normal) > math.pi / 2:  # reversed normal
```

Comment:
Yes, that is true and we can add an angle_tolerance input but I know that it will get annoying for people working with the Python API to have to specify an angle tolerance every time they want to add an aperture.  If this were a case where we could use a default tolerance of 0  and everything would usually be ok, I would do that but most faces that aren't perfectly vertical or horizontal will raise an exception in this case simply because of Python floating point tolerance.  I'm also not sure that adding an angle_tolerance check here will really add much more efficiency to the checks since you probably want to run the `check_are_sub_faces_valid()` anyway in order to verify that the sub-faces lie within the boundary of the parent face.

Reply:
That was not what I meant but I see that what you are trying to do here is different then ensuring the faces are on the same plane so it should be good.

I think on the model, room and face we should have `validate` methods that does all the checks. It doesn't need to run by default but it is an option for to ensure model quality.

Reply:
If you don't mind, I'm just going to keep this unresolved since I agree that we can expose methods on room and model that check all children faces.  It's a simple thing to add given that the Face, Aperture, and Door, already have the check methods so I can do it in this PR.

Reply:
Done!  I added in this capability on both the Room and the Model.